File: python/2300.Successful-Pairs-of-Spells-and-Potions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        result = []
        potions.sort()

        for spell in spells:
            left = 0
            right = len(potions) - 1
            index = len(potions)

            while left <= right:
                mid = (left + right) // 2
                potion_success = spell * potions[mid]

                if potion_success >= success:
                    index = mid
                    right = mid - 1
                else:
                    left = mid + 1
            
            result.append(len(potions) - index)

        return result

File: javascript/2300.Successful-Pairs-of-Spells-and-Potions.js
/**
 * @param {number[]} spells
 * @param {number[]} potions
 * @param {number} success
 * @return {number[]}
 */
var successfulPairs = function (spells, potions, success) {
  const result = [];
  potions.sort((a, b) => a - b);

  for (const spell of spells) {
    let index = potions.length;
    let left = 0;
    let right = potions.length - 1;

    while (left <= right) {
      const mid = Math.floor((left + right) / 2);
      const potion_success = spell * potions[mid];

      if (potion_success >= success) {
        right = mid - 1;
        index = mid;
      } else {
        left = mid + 1;
      }
    }

    result.push(potions.length - index);
  }

  return result;
};
